staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/dispatcher.go

```diff
@@ -100,11 +100,25 @@ func (d *validatingDispatcher) Dispatch(ctx context.Context, attr admission.Attr
 	}
 
 	wg := sync.WaitGroup{}
-	errCh := make(chan error, len(relevantHooks))
+	errCh := make(chan error, 2*len(relevantHooks)) // double the length to handle extra errors for panics in the gofunc
 	wg.Add(len(relevantHooks))
 	for i := range relevantHooks {
 		go func(invocation *generic.WebhookInvocation, idx int) {
-			defer wg.Done()
+			ignoreClientCallFailures := false
+			defer func() {
+				defer wg.Done()
```

Comment:
defer inside defer is really hard to follow... is there a reason not to keep the `defer wg.Done()` where it was?

Reply:
> defer inside defer is really hard to follow... is there a reason not to keep the defer wg.Done() where it was?

Calling Done of the wait group causes the parent thread to progress and close the error channel that this defer is attempting to write to.  This causes a race between the defer go func and the parent go func over which happens first.

If memory serves, it is possible to order the defer statement and gain a guaranteed happens-before, but that seemed more likely to be prone to future refactoring failures. I'm willing to structure it either way or add a comment since the reasoning is not obvious on a top to bottom read.

Reply:
When you get it wrong `go test k8s.io/apiserver/pkg/admission/plugin/webhook/validating -run "TestValidatePanicHandling" --race` produces

> --- FAIL: TestValidatePanicHandling (0.32s)
    testing.go:1152: race detected during execution of test

But I haven't figured out how to merge the tests and depending on, "be careful to add these defers in the right order" wasn't clearly better to me than "defer inside of defer ensures correct ordering".  I'm flexible about factorization if your preference is to rely on the order in which defers are added, which is apparently called in reverse order (not the forward order that I would naively expect).

Reply:
non-nested defers called in reverse order reads more easily to me